drivers/hmis/app/graphql/mutations/ce/submit_ce_referral_step.rb

```diff
@@ -16,23 +16,65 @@ class Ce::SubmitCeReferralStep < CleanBaseMutation
     argument :referral_id, ID, required: true
     argument :step_id, ID, required: true
     argument :input, Types::JsonObject, required: true
-    field :step, Types::HmisSchema::CeReferralStep, null: false
-    field :referral, Types::HmisSchema::CeReferral, null: false # return the referral so the UI can respond appropriately if the referral's overall status has changed
+    argument :confirmed, Boolean, required: false
 
-    def resolve(referral_id:, step_id:, input:)
+    field :step, Types::HmisSchema::CeReferralStep, null: true # nullable in case of errors
+    field :referral, Types::HmisSchema::CeReferral, null: true # return the referral so the UI can respond appropriately if the referral's overall status has changed
+
+    def resolve(referral_id:, step_id:, input:, confirmed: false)
       raise unless Hmis::Ce.configuration.enabled?
 
       referral = Hmis::Ce::Referral.viewable_by(current_user).find(referral_id)
-      step = nil
+      errors = HmisErrors::Errors.new
+
+      unless confirmed
+        # dependency injection style engine
+        dry_run_engine = Hmis::WorkflowExecution::Engine.new(
+          referral.workflow_instance,
+          message_handler: Hmis::Ce::DryRunMessageHandler.new,
+          stepper: Hmis::Ce::DryRunEngineStepper.new,
+          assignment_handler: nil, # dry runner does not do any assignment
+          audit_logger: nil, # or log audit events
+        )
+
+        step = dry_run_engine.active_steps.find(step_id)
+
+        errors.push(*validate(step, input))
+        return { errors: errors } if errors.any? # return early - if there are validation errors, they may cause the dry run to fail
+
+        dry_run_engine.complete_step!(step, user: current_user, submitted_values: input)
+
+        # todo @martha - this isn't flexible enough, it needs to be configurable which types of messages require a warning.
```

Comment:
I'd like to think this through together before proceeding. 

Currently, the MessageHandler only handles events happening to the whole _referral_ (like start referral, reject referral, etc.) But it doesn't handle events happening on individual steps, like start step, complete step.

That seems easy enough to address, maybe using the new Stepper I've introduced here, to collect messages similarly to how DryRunMessageHandler does. However, once we do, we still need some additional configuration, maybe a new boolean column on Nodes, indicating whether starting a given node is a warning-level event.

Does that sound right? Does it sound worth it? Or should we just say that only referral events can trigger warnings, not step events?

Reply:
Thanks Theron, this code sketch is really helpful! I like this idea. Let's discuss at dev standup, since I also agree with
> It may not be worth if we could make do with warnings or guidance inlined into the steps forms using existing form definitions.

My recollection from our design meetings is that:
- We first added the dry-run/confirm warnings idea based on the idea that configurable forms means entrusting any warning elements for irreversible actions to end users, and we want to avoid that
- However, after more discussion with design, we agreed that having a warning in the form (pre-submit) is a requirement, and we may add new form elements to support this(?)
- so we may have elided the need for an infinitely configurable warning confirmation?

Reply:
Discussion notes from dev standup
- With the current approach, we could just use the `send_notification` event rather than notifying on step change
- But we would also want this to be configurable per referral template. Not all templates want to warn/confirm on every email notification
- In the future let's consider building confirmation into the trigger config
- For now let's keep the API changes with the `confirmed` flag, but revert the dry run engine since it adds unneeded complexity
- Move validation logic into the step itself, later this will be flexible to return warnings as well as errors